packages/warehouses/src/ssh/sshTunnel.ts

```diff
@@ -0,0 +1,82 @@
+import SSH2Promise from 'ssh2-promise';
+import SSHConfig from 'ssh2-promise/lib/sshConfig';
+
+export class SshTunnel<T extends CreateWarehouseCredentials> {
+    readonly originalCredentials: T;
+
+    overrideCredentials: T;
+
+    localPort: number | undefined;
+
+    private sshConnection: SSH2Promise | undefined;
+
+    constructor(credentials: T) {
+        this.originalCredentials = credentials;
+        this.overrideCredentials = credentials;
+        this.localPort = undefined;
+        this.sshConnection = undefined;
+    }
+
+    connect = async (): Promise<T> => {
+        const { type } = this.originalCredentials;
+        switch (type) {
+            case WarehouseTypes.POSTGRES:
+            case WarehouseTypes.REDSHIFT:
+                if (this.originalCredentials.useSshTunnel) {
+                    try {
+                        const remoteHostConfig = {
+                            host: this.originalCredentials.sshTunnelHost || '',
+                            port: this.originalCredentials.sshTunnelPort || 22,
+                            username:
+                                this.originalCredentials.sshTunnelUser || '',
+                            privateKey:
+                                this.originalCredentials.sshTunnelPrivateKey ||
+                                '',
+                        } as SSHConfig;
+                        this.sshConnection = new SSH2Promise(remoteHostConfig);
+                        console.info('Opening SSH tunnel to remote host');
+                        const sshTunnel = await this.sshConnection.addTunnel({
+                            remoteAddr: this.originalCredentials.host,
+                            remotePort: this.originalCredentials.port,
+                        });
+                        console.info(
+                            `SSH tunnel ready at ${sshTunnel.localPort}`,
+                        );
+                        this.localPort = sshTunnel.localPort;
+                        this.overrideCredentials = {
+                            ...this.originalCredentials,
+                            host: '127.0.0.1',
+                            port: sshTunnel.localPort,
+                        };
+                    } catch (e) {
+                        throw new WarehouseConnectionError(
+                            `Could not open SSH tunnel: ${e.message}`,
+                        );
+                    }
+                }
+                break;
+            case WarehouseTypes.DATABRICKS:
+            case WarehouseTypes.TRINO:
+            case WarehouseTypes.SNOWFLAKE:
+            case WarehouseTypes.BIGQUERY:
+                break;
+            default:
+                assertUnreachable(type, new Error('Unknown warehouse type'));
+        }
+        return this.overrideCredentials;
+    };
+
+    disconnect = async (): Promise<void> => {
+        if (this.sshConnection) {
+            console.info('Closing SSH tunnel');
```

Comment:
suggestion: we should use our `logger.debug` instead of `console.info` to match the rest of the logs.

Reply:
I started with this but I don't think we make the logger available in the warehouses package? Is there a way to do that so it integrates well with the logger in the main server?